src/static/js/chapter.js

```diff
@@ -0,0 +1,127 @@
+    const cvs = document ? document.createElement('canvas') : null;
+    const ctx = cvs && cvs.getContext ? cvs.getContext('2d') : null;
+    cvs.width = 6000;
+    cvs.height = 3700;
+    ctx.fillRect(5999, 3699, 1, 1);
+    largeCanvasSupported = Boolean(ctx.getImageData(5999, 3699, 1, 1).data[3]) == 1;
+  }
+  catch (e) {
+    largeCanvasSupported = false;
+  }
+
+  if (!largeCanvasSupported) {
+    console.log('High resolution canvas images are not supported');
+  }
+
+  return largeCanvasSupported;
+
+};
+
+//Google Sheets is sometimes blocked by companies/proxies so let's check for that
+//Unfortunately no easy way to check this as cannot make non-CORS request to
+//Another domain and can't check status response when CORS domain :-(
+//Below seems to work for my corporate proxy anyway
+const googleSheetsAllowed = async () => {
+
+  let sheetsAllowed = true;
+
+  const url = 'https://docs.google.com/spreadsheets/';
+
+  //non-async aware browsers can stick with images, as they probably don't support sheet anyway
+  await fetch(url, { method: 'HEAD', mode: 'no-cors' })
+    .then(
+      function (response) {
+        //If response status is 0 then means CORS is blocking it
+        //Which means it made it as proxies block with a 403
+        //before getting that far.
+        if (response.status != 0) {
+          sheetsAllowed = false;
+        }
+      }
+    )
+    .catch(function (err) {
+      sheetsAllowed = false;
+    });
+
+  if (!sheetsAllowed) {
+    console.log('Google Sheets access blocked');
+  }
+  return sheetsAllowed;
+}
+
+//We use Google Sheets for detailed visualisations
+//Check for support and switch out images if supported
+const upgradeInteractiveFigures = async () => {
+
+  try {
+    if (highResolutionCanvasSupported() && !dataSaverEnabled() && await googleSheetsAllowed()) {
+      console.log('Upgrading to interactive figures');
+
+      //Set the Google Sheets iframe
+      let all_fig_imgs = document.querySelectorAll('figure img');
+
+      all_fig_imgs.forEach(function (fig_img) {
+
+        if (fig_img.getAttribute('data-src')) {
+
+          var iframe = document.createElement('iframe');
+
+          //Set up some default attributes
+          iframe.setAttribute('aria-describedby', fig_img.getAttribute('aria-describedby'));
+          iframe.setAttribute('title', fig_img.getAttribute('alt'));
+          iframe.setAttribute('width', "600");
+          iframe.setAttribute('height', '371');
+          iframe.setAttribute('seamless', '');
+          iframe.setAttribute('frameborder', '0');
+          iframe.setAttribute('scrolling', 'no');
+          iframe.setAttribute('loading', 'lazy');
+
+          //Copy the atttributes from the data- attributes of the img
+          for (var i = 0, atts = fig_img.attributes, n = atts.length, arr = []; i < n; i++) {
+            const att_name = atts[i].nodeName;
+            if (att_name.substring(0, 5) == 'data-') {

```

Comment:
Can't we use the dataset API directly? There are only a few attributes we would use so we can look them up individually (in a loop) as needed.

Reply:
TIL! Never heard of that before. It is cleaner - though does mean we've a hardcoded list but, as you say, shouldn't change.